pkg/terraform/store.go

```diff
@@ -0,0 +1,151 @@
+type Setup struct {
+	Version       string
+	Requirement   ProviderRequirement
+	Configuration ProviderConfiguration
+}
+
+func (p Setup) validate() error {
+	if p.Version == "" {
+		return errors.New(fmtErrValidationVersion)
+	}
+	if p.Requirement.Source == "" || p.Requirement.Version == "" {
+		return errors.Errorf(fmtErrValidationProvider, p.Requirement.Source, p.Requirement.Version)
+	}
+	return nil
+}
+
+func NewWorkspaceStore() *WorkspaceStore {
+	return &WorkspaceStore{store: map[types.UID]*Workspace{}, mu: sync.Mutex{}}
+}
+
+type WorkspaceStore struct {
+	// store holds information about ongoing operations of given resource.
+	// Since there can be multiple calls that add/remove values from the map at
+	// the same time, it has to be safe for concurrency since those operations
+	// cause rehashing in some cases.
+	store map[types.UID]*Workspace
+	mu    sync.Mutex
+}
+
+// TODO(muvaf): Take EnqueueFn as parameter tow WorkspaceStore?
+
+func (ws *WorkspaceStore) Workspace(ctx context.Context, tr resource.Terraformed, ts Setup, l logging.Logger, _ EnqueueFn) (*Workspace, error) {
+	dir := filepath.Join(os.TempDir(), string(tr.GetUID()))
+	fp, err := NewFileProducer(tr, ts)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create a new file producer")
+	}
+	if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+		return nil, errors.Wrap(err, "cannot create directory for workspace")
+	}
+	_, err = os.Stat(filepath.Join(dir, "terraform.tfstate"))
+	if xpresource.Ignore(os.IsNotExist, err) != nil {
+		return nil, errors.Wrap(err, "cannot state terraform.tfstate file")
+	}
+	if os.IsNotExist(err) {
+		s, err := fp.TFState()
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot produce tfstate")
+		}
+		rawState, err := json.JSParser.Marshal(s)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot marshal state object")
+		}
+		if err := os.WriteFile(filepath.Join(dir, "terraform.tfstate"), rawState, os.ModePerm); err != nil {
+			return nil, errors.Wrap(err, "cannot write tfstate file")
+		}
+	}
+	rawHCL, err := json.JSParser.Marshal(fp.MainTF())
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot marshal main hcl object")
+	}
+	if err := os.WriteFile(filepath.Join(dir, "main.tf.json"), rawHCL, os.ModePerm); err != nil {
+		return nil, errors.Wrap(err, "cannot write tfstate file")
+	}
+	// TODO(muvaf): Set new logger every time?
+	ws.mu.Lock()
+	w, ok := ws.store[tr.GetUID()]
+	if !ok {
+		ws.store[tr.GetUID()] = NewWorkspace(dir, WithLogger(l))
+		w = ws.store[tr.GetUID()]
+	}
+	ws.mu.Unlock()
+	// The operation has ended but there could be a lock file if Terraform crashed.
+	// So, we clean that up here. We could check the existence of the file first
+	// but just deleting in all cases get us to the same state with less logic and
+	// same number of fs calls.
+	if w.LastOperation.EndTime != nil {
+		if err := os.RemoveAll(filepath.Join(dir, ".terraform.tfstate.lock.info")); err != nil {
```

Comment:
I just tried using `force-unlock` but while I'm doing that I realized there is [`-lock=false`](https://www.terraform.io/docs/cli/commands/apply.html#lock-false) flag where we can disable creation of a lock. It sounds dangerous at first but from my reading it's intended to prevent concurrent operations and we never do that, so we should be fine with using the flag.

What I did was to add a different check for `.terraform.lock.hcl` in order to skip `terraform init`, which doesn't take too much time but still much more than a simple `os.Stat` call.